jeff65/gold/lexer.py

```diff
@@ -210,13 +212,18 @@ def lex(stream):
                 position, scan_word(source, c), ast.IdentifierNode)
 
         # nestable delimiters (parentheses)
-        elif c in nestable_delimiters:
+        elif comment_depth is 0 and c in nestable_delimiters:
             yield make_token(position, c)
 
         # non-alphanumeric word (mostly operators)
         # we call these "sprinkles"
         else:
-            yield make_token(position, scan_sprinkle(source, c))
+            t = make_token(position, scan_sprinkle(source, c))
```

Comment:
keeping track of comment nesting depth in the lexer doesn't seem like an optimal solution, but it's the best one that has occurred to me so far.

@jdpage do you have any ideas? i would really rather not use parentheses on the right-hand side of an array declaration.

Reply:
actually, we have another problem. say i have:

    let x: [u8; 2, 2] = [[0, 1], [2, 3]]

and i want to comment it out:

    --[[ let x: [u8; 2, 2] = [[0, 1], [2, 3]] ]]

that's a problem.

i think we have no choice except to either use parentheses on the right of the assignment or to change the syntax for comments.

i don't like parentheses on the right because parentheses are already used for multiple things.

Reply:
i know you think they're ugly, but now we have a perfect example of why so many languages use `/*` and `*/`. we need a better argument than "they're ugly".

Reply:
I actually don't see the problem; `--[[x=values[offsets[y]]]]` is unambiguous. Just don't think of the `]]` comment-ender as a single token; it's really two tokens, and the comment-opener is looking for two consecutive `]` tokens to match instead of a single `]]` token.

Reply:
i am assuming comments should follow these syntax rules (from the user point of view):

1. a comment can contain any text without breaking the parser, except for the "open comment" and "end comment" token.
2. there must be exactly one "close comment" token for each "open comment" token. this supports nesting.

let's consider the naive implementation of parsing `--[[x=values[offsets[y]]]]`. it would result in a comment containing `x=values[offsets[y` and two unmatched close brackets left in the stream (because the comment ends at the first instance of double-close-bracket). this is obviously wrong.

a "smarter" way of parsing that would be:

1. the `--` token creates a `CommentNode`. all of the square brackets would just create square bracket nodes (open or close as appropriate).
2. the `eat_comment` function of `CommentNode` would:
   1. make sure that the next two tokens in the stream are open bracket nodes
   2. set `depth` to 2.
   3. eat right tokens until depth is zero. each open bracket increments depth, each close bracket decrements depth.
   4. do something to make sure that the last two brackets were adjacent.

these rules would properly parse `--[[x=values[offsets[y]]]]`; however, it violates rule 1. i can no longer put unmatched brackets inside of a comment. this is unexpected and violates a pretty core concept of "comments". i consider that unacceptable.

Reply:
another option is using bare strings as comments, like in python.